### PR TITLE
Add `FusionDefinition` methods to print the `fusion_ir`, `scheduled_fusion_ir`, and `cuda_code`

### PR DESCRIPTION
**Pull request by kevinstephano**
Resolves issue #2387 for @mruberry .

The following APIs are added to the Python `FusionDefinition`.  They all return strings instead of printing to `stdout` so the python user can do what they want with the string output.

```python
fd.fusion_ir()
fd.last_scheduled_fusion_ir(tensor_transforms=False)
fd.scheduled_fusion_ir_for(inputs, tensor_transforms=False)
fd.last_cuda_code(intrinsic_code=False)
fd.cuda_code_for(inputs, intrinsic_code=False)
```

Example test case:

```python
import torch
from nvfuser import FusionDefinition

inputs = [
    torch.randn(4, 4, device='cuda'),
]

with FusionDefinition() as fd:
    t0 = fd.from_pytorch(inputs[0])
    t1 = fd.ops.relu(t0)
    fd.add_output(t1)

out = fd.execute(inputs)
```

**Fusion IR**
Gives a string for the unscheduled Fusion IR for a given definition.
Example Usage:
```python
print(fd.fusion_ir())
```
Output:
```
%kernel {
T1_g[ iS2{i0}, iS3{i1} ]
   = relu(T0_g[ iS0{i0}, iS1{i1} ]);
}
```

**Scheduled Fusion IR**
Gives a string for the scheduled Fusion IR with or without transforms for either the last scheduled fusion or for a specific set of inputs.  Note, segmented fusions also print the segmented groupings to help the user understand what the scheduled Fusion IR represents.

Example Usages:
```python
print(fd.last_scheduled_fusion_ir(tensor_transforms=True))
print(fd.scheduled_fusion_ir_for(inputs, tensor_transforms=True))
```
Output:
```
%kernel {
T2_l[ iblockIdx.x27{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS28{1}, iS26{1}, ithreadIdx.x24{128} ] ca_pos( 2 )
   = T0_g[ iS34{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iS35{1}, iS33{1}, iS31{128} ];
T3_l[ iblockIdx.x20{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS21{1}, iS19{1}, ithreadIdx.x17{128} ] ca_pos( 4 ) produce_pos( 2 )
   = relu(T2_l[ iblockIdx.x27{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS28{1}, iS26{1}, ithreadIdx.x24{128} ] ca_pos( 2 ));
T1_g[ iblockIdx.x13{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS14{1}, iS12{1}, ithreadIdx.x10{128} ] ca_pos( 2 ) produce_pos( 4 )
   = T3_l[ iblockIdx.x20{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS21{1}, iS19{1}, ithreadIdx.x17{128} ] ca_pos( 4 ) produce_pos( 2 );

TransformPrinter : 
T0_g[ iS34{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iS35{1}, iS33{1}, iS31{128} ]
 root domain : (iS36{T0.size[0]},iS37{T0.size[1]})
 contiguity: t t
  Merge: iS36{T0.size[0]} and iS37{T0.size[1]} -> iS29{( T0.size[0] * T0.size[1] )}
  Split: iS29{( T0.size[0] * T0.size[1] )} by factor 128 -> iS30{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )}, iS31{128}, start offset: 0, stop offset: 0
  Split: iS30{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )} by factor 1 -> iS32{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )}, iS33{1}, start offset: 0, stop offset: 0
  Split: iS32{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )} by factor 1 -> iS34{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iS35{1}, start offset: 0, stop offset: 0
T2_l[ iblockIdx.x27{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS28{1}, iS26{1}, ithreadIdx.x24{128} ] ca_pos( 2 )
 root domain : (iS38{T0.size[0]},iS39{T0.size[1]})
 contiguity: t t
  Merge: iS38{T0.size[0]} and iS39{T0.size[1]} -> iS22{( T0.size[0] * T0.size[1] )}
  Split: iS22{( T0.size[0] * T0.size[1] )} by factor 128 -> iS23{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )}, ithreadIdx.x24{128}, start offset: 0, stop offset: 0
  Split: iS23{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )} by factor 1 -> iS25{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )}, iS26{1}, start offset: 0, stop offset: 0
  Split: iS25{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )} by factor 1 -> iblockIdx.x27{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS28{1}, start offset: 0, stop offset: 0
T3_l[ iblockIdx.x20{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS21{1}, iS19{1}, ithreadIdx.x17{128} ] ca_pos( 4 ) produce_pos( 2 )
 root domain : (iS40{T0.size[0]},iS41{T0.size[1]})
 contiguity: t t
  Merge: iS40{T0.size[0]} and iS41{T0.size[1]} -> iS15{( T0.size[0] * T0.size[1] )}
  Split: iS15{( T0.size[0] * T0.size[1] )} by factor 128 -> iS16{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )}, ithreadIdx.x17{128}, start offset: 0, stop offset: 0
  Split: iS16{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )} by factor 1 -> iS18{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )}, iS19{1}, start offset: 0, stop offset: 0
  Split: iS18{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )} by factor 1 -> iblockIdx.x20{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS21{1}, start offset: 0, stop offset: 0
T1_g[ iblockIdx.x13{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS14{1}, iS12{1}, ithreadIdx.x10{128} ] ca_pos( 2 ) produce_pos( 4 )
 root domain : (iS42{T0.size[0]},iS43{T0.size[1]})
 contiguity: t t
  Merge: iS42{T0.size[0]} and iS43{T0.size[1]} -> iS8{( T0.size[0] * T0.size[1] )}
  Split: iS8{( T0.size[0] * T0.size[1] )} by factor 128 -> iS9{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )}, ithreadIdx.x10{128}, start offset: 0, stop offset: 0
  Split: iS9{( ceilDiv(( T0.size[0] * T0.size[1] ), 128) )} by factor 1 -> iS11{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )}, iS12{1}, start offset: 0, stop offset: 0
  Split: iS11{( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) )} by factor 1 -> iblockIdx.x13{( ceilDiv(( ceilDiv(( ceilDiv(( T0.size[0] * T0.size[1] ), 128) ), 1) ), 1) )}, iUS14{1}, start offset: 0, stop offset: 0
}
```

**Cuda Code**
Gives a string for the Cuda Code for a fusion with or without the full intrinsic code required to execute the kernel for either the last scheduled fusion or for a specific set of inputs.

Example Usages:
```python
print(fd.last_cuda_code(intrinsic_code=False))
print(fd.cuda_code_for(inputs, intrinsic_code=False))
```

Output:
```c++
__global__ void kernel1(Tensor<float, 2> T0, Tensor<float, 2> T1) {
  int i50;
  i50 = ((nvfuser_index_t)threadIdx.x) + (128 * ((nvfuser_index_t)blockIdx.x));
  if ((i50 < (T0.size[0] * T0.size[1]))) {
    float T2[1];
    T2[0] = 0;
    T2[0]
       = T0[i50];
    float T3[1];
    T3[0]
       = relu(T2[0]);
    T1[i50]
       = T3[0];
  }
}
```

### CODE REVIEW

File: third_party/nvfuser/csrc/executor.h [hunk 1]
```diff
@@ -241,6 +246,10 @@ class TORCH_CUDA_CU_API FusionExecutor : public NonCopyable {
   std::vector<at::Tensor> allocOutputSpace(
       const at::ArrayRef<c10::IValue>& inputs);
 
+  Fusion* fusion() const {
```
Comment (naoyam) [inline third_party/nvfuser/csrc/executor.h:249]: Doesn't this return the same as `kernel()`? Perhaps, we should remove `fusion_` and replace it with `kernel()`.

File: third_party/nvfuser/csrc/kernel_cache.cpp [hunk 2]
```diff
@@ -222,6 +222,78 @@ std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
...
+    } else {
+      kernel_code += "\n";
+    }
+    kernel_code += exec.kernelString();
+  }
+
+  if (intrinsic_code) {
+    const FusionExecutor& fe = kernel_runtime->executors()[0];
+    std::string full_code =
+        fe.getStructuredCode(kernel_code, fe.kernel()->indexType());
```
Comment (naoyam) [inline third_party/nvfuser/csrc/kernel_cache.cpp:245]: Do all executors have the same index type? Even though I recently did some cleanup of the kernel index type, I'm still unsure here.

Comment (kevinstephano) [reply third_party/nvfuser/csrc/kernel_cache.cpp:245]: I assumed the code I am calling the "intrinsic_code" which we print as the "full_kernel" is identical for each kernel in segmentation.  Is that an incorrect assumption?  I could also print the "full_kernel" for each kernel of segmentation but I thought that might be too verbose.

Comment (naoyam) [reply third_party/nvfuser/csrc/kernel_cache.cpp:245]: `getStructuredCode` inserts a macro definition of the index type, which is either `int` or `int64_t`. The index type of a FusionExecutor is determined by the KernelArgumentHolder passed to `compileFusion`. I think it only holds the inputs to each segment of the original fusion, so it seems possible to have different segments to have different index types.

Comment (naoyam) [reply third_party/nvfuser/csrc/kernel_cache.cpp:245]: For example, if only one segment uses a very large tensor that exceeds the 32-bit indexing limit, only that segment would have the index type defined as `int64_t`, and the rest of the segments should use `int`. I haven't verified this behavior, though.

Comment (kevinstephano) [reply third_party/nvfuser/csrc/kernel_cache.cpp:245]: It seems like you are advocating for printing the "intrinsic code" for each segment given that the `indexType()` could change per segment.

Comment (naoyam) [reply third_party/nvfuser/csrc/kernel_cache.cpp:245]: I'm just saying the printed code done in this PR may not be correct. Maybe it doesn't matter? Not sure what is expected.

**Comment by kevinstephano** [comment]
Closing in favor of PR in new repo: https://github.com/NVIDIA/Fuser/pull/6

**Pull request closed without merging**